### PR TITLE
Share the TritonTF model across instances on same device

### PR DESCRIPTION
**Pull request by GuanLuo**
However, I don't see any performance improvement:
config
```
cat models/resnet_v1_50_savedmodel/config.pbtxt 
name: "resnet_v1_50_savedmodel"
platform: "tensorflow_savedmodel"
max_batch_size: 128
input [
  {
    name: "input"
    data_type: TYPE_FP32
    format: FORMAT_NHWC
    dims: [ 224, 224, 3 ]
  }
]
output [
  {
    name: "output"
    data_type: TYPE_FP32
    dims: [ 1000 ]
    label_filename: "resnet50_labels.txt"
  }
]
instance_group [ {count: 4} ]
```
before
```
qa/clients/perf_analyzer -m resnet_v1_50_savedmodel -b 128 --concurrency-range 4:4:1
*** Measurement Settings ***
  Batch size: 128
  Measurement window: 5000 msec
  Latency limit: 0 msec
  Concurrency limit: 4 concurrent requests
  Using synchronous calls for inference
  Stabilizing using average latency

Request concurrency: 4
  Client: 
    Request count: 46
    Throughput: 1177.6 infer/sec
    Avg latency: 443769 usec (standard deviation 33161 usec)
    p50 latency: 451443 usec
    p90 latency: 470631 usec
    p95 latency: 478158 usec
    p99 latency: 541150 usec
    Avg HTTP time: 445176 usec (send/recv 133191 usec + response wait 311985 usec)
  Server: 
    Inference count: 6912
    Execution count: 54
    Successful request count: 54
    Avg request latency: 296588 usec (overhead 48 usec + queue 54 usec + compute input 65821 usec + compute infer 230548 usec + compute output 117 usec)

Inferences/Second vs. Client Average Batch Latency
Concurrency: 4, throughput: 1177.6 infer/sec, latency 443769 usec
```
after
```
qa/clients/perf_analyzer -m resnet_v1_50_savedmodel -b 128 --concurrency-range 4:4:1
*** Measurement Settings ***
  Batch size: 128
  Measurement window: 5000 msec
  Latency limit: 0 msec
  Concurrency limit: 4 concurrent requests
  Using synchronous calls for inference
  Stabilizing using average latency

Request concurrency: 4
  Client: 
    Request count: 45
    Throughput: 1152 infer/sec
    Avg latency: 432710 usec (standard deviation 36968 usec)
    p50 latency: 435244 usec
    p90 latency: 468676 usec
    p95 latency: 485366 usec
    p99 latency: 544514 usec
    Avg HTTP time: 431544 usec (send/recv 131653 usec + response wait 299891 usec)
  Server: 
    Inference count: 7296
    Execution count: 57
    Successful request count: 57
    Avg request latency: 283689 usec (overhead 47 usec + queue 54 usec + compute input 62902 usec + compute infer 220574 usec + compute output 112 usec)

Inferences/Second vs. Client Average Batch Latency
Concurrency: 4, throughput: 1152 infer/sec, latency 432710 usec
```

### CODE REVIEW

**Comment by CoderHam** [comment]
While the performance look similar, how is the memory consumption affected if we run 2 model instances on a single GPU?

**Review by tanmayv25 — CHANGES_REQUESTED** [review]

**Comment by tanmayv25** [comment]
This PR will allow us to set higher instance_count for TensorFlow as the sessions will be shared. The performance with the same instance count is expected to be within similar range as there will be contention over the resources. But having a parameter to control how many instances to share a session will be a useful tuning knob. This should tie in well with `TF_USE_PER_SESSION_THREADS` knob exposed in PR #15 to allay some of the CPU side contention with per session thread pools.

File: src/tensorflow.cc [hunk 1]
```diff
@@ -698,8 +723,228 @@ class ModelState : public BackendModel {
...
+    }
+    RETURN_IF_TRITONTF_ERROR(TRITONTF_ModelMakeCallable(
+        lmodel_handle.get(), input_names.data(), input_types.data(),
+        config_inputs.ArraySize(), output_names.data(), output_types.data(),
+        config_outputs.ArraySize()));
+  }
+
+  lmodel.tritontf_model_ = lmodel_handle.get();
+  model_handles_.emplace_back(std::move(lmodel_handle));
+  models_[device_id] = std::make_pair(1, lmodel);
```
Comment (tanmayv25) [inline src/tensorflow.cc:943]: So, after every `max_session_share_count_` (and first session creation) the models_[device_id] is updated to 1. So, the counting can begin anew. The logic looks little separated from top. Will be better to include `Model lmodel` initialization(L743 to L939) into separate function to make this easier to read and understand.

**Review by tanmayv25 — CHANGES_REQUESTED** [review]

File: src/tensorflow.cc (hunk 1 above)
Comment (GuanLuo) [reply src/tensorflow.cc:943]: This is already the function for initializing the model. Other than that, there is only the top and the bottom where it updates the corresponding members in ModelState

Comment (tanmayv25) [reply src/tensorflow.cc:943]: Rename the current CreateModel to GetModel and include the section I pointed into a new abstraction called CreateModel().

**Review by tanmayv25 — APPROVED** [review]

**Pull request merged**